### PR TITLE
Reduce gulcalc memory use for gul alloc rule 1

### PR DESCRIPTION
**Pull request by hchagani-oasislmf**
In previous version, calculation of losses performed as `getmodel` stream read in. This resulted in holding the losses in memory until exposure value could be calculated. By moving calculation of losses to follow the determination of number of items per coverage ID per event ID (necessary to determine exposure value as `TIV / number_of_items`), these values no longer need to be stored in memory for each event ID. Rather, these values can be stored in memory per coverage ID.

To facilitate this, the calculated mean and standard deviation, item ID and group ID for each entry from `getmodel` is stored in memory. Additionally, a map of (`prob_to`, `bin_mean`) tuples to bin IDs are established in order to create a lookup table, thus saving memory.

Additionally, a benchmark option (instigated with `-b` argument on command line) has been introduced to analyse memory use by certain global variables. This is still in development.

### CODE REVIEW

File: src/gulcalc/gulcalc.cpp [hunk 1]
```diff
@@ -133,100 +227,204 @@ void splittiv(std::vector<gulItemIDLoss> &gulitems, OASIS_FLOAT tiv)
...
 }
-void gulcalc::clearfullCorr_data()
-{
-	auto iter = fullCorr_.begin();
-	while (iter != fullCorr_.end()) {
-		auto iter2 = iter->begin();
-		while (iter2 != iter->end()) {
-			iter2->clear();
+
+void gulcalc::WriteMode1Output(const int event_id, const OASIS_FLOAT tiv,
```
Comment (ndaqjn) [inline src/gulcalc/gulcalc.cpp:242]: Use the same casing rules for all the code in one class

File: src/gulcalc/gulcalc.cpp [hunk 2]
```diff
@@ -80,10 +80,104 @@ bool operator<(const gulitemSampleslevel& lhs, const gulitemSampleslevel& rhs)
...
+	}
+
+}
+
 gulSampleslevelHeader lastitemheader;
 gulSampleslevelHeader lastcorrelatedheader;
 gulSampleslevelHeader lastcoverageheader;
 
+// Benchmark memory usage
+void gulcalc::OutputBenchmark(const int event_id, bool header=false) {
```
Comment (ndaqjn) [inline src/gulcalc/gulcalc.cpp:99]: Use the same casing

File: src/gulcalc/gulcalc.cpp [hunk 3]
```diff
@@ -258,66 +456,6 @@ void gulcalc::outputcoveragedata(int event_id)
 	cov_.resize(coverages_->size());
```
Comment (ndaqjn) [inline src/gulcalc/gulcalc.cpp:456]: If the size is smaller than the current size, should this step be skipped?

File: src/gulcalc/gulcalc.cpp [hunk 4]
```diff
@@ -798,6 +885,8 @@ void gulcalc::mode1()
 	char* rec = new char[max_recsize];
 	damagecdfrec* d = (damagecdfrec*)rec;
 
+	if (benchmark_) OutputBenchmark(0, true);
+
 	for (;;)
```
Comment (ndaqjn) [inline src/gulcalc/gulcalc.cpp:890]: What makes it exit this loop?

File: src/gulcalc/gulcalc.cpp [hunk 5]
```diff
@@ -843,6 +932,8 @@ void gulcalc::mode0()
 	
 	damagecdfrec *d = (damagecdfrec *)rec;
 
+	if (benchmark_) OutputBenchmark(0, true);
+
 	for (;;)
```
Comment (ndaqjn) [inline src/gulcalc/gulcalc.cpp:937]: Same here, what makes it exit the loop?

File: src/gulcalc/gulcalc.cpp (hunk 3 above)
Comment (hchagani-oasislmf) [reply src/gulcalc/gulcalc.cpp:456]: I'm not sure. This is used for a deprecated output stream. I can take a look but considering this PR forms part of a patch, I would prefer to submit this PR and create another issue regarding the above. Would you agree?

Comment (ndaqjn) [reply src/gulcalc/gulcalc.cpp:456]: If it's deprecated, then I agree. Leave it as is.

File: src/gulcalc/gulcalc.cpp (hunk 4 above)
Comment (hchagani-oasislmf) [reply src/gulcalc/gulcalc.cpp:890]: When the input stream ends. See L893-894:
```
                bool bSuccess = iGetrec_(p, sizeof(damagecdfrec));
                if (bSuccess == false) break;
```

File: src/gulcalc/gulcalc.cpp (hunk 5 above)
Comment (hchagani-oasislmf) [reply src/gulcalc/gulcalc.cpp:937]: As above, this time see L940-941:
```
                bool bSuccess = iGetrec_(p, sizeof(damagecdfrec));
                if (bSuccess == false) break;
```

File: src/gulcalc/gulcalc.cpp (hunk 1 above)
Comment (hchagani-oasislmf) [reply src/gulcalc/gulcalc.cpp:242]: Understood; will change this.

File: src/gulcalc/gulcalc.cpp (hunk 2 above)
Comment (hchagani-oasislmf) [reply src/gulcalc/gulcalc.cpp:99]: As above, will change this.

File: src/gulcalc/gulcalc.cpp (hunk 4 above)
Comment (ndaqjn) [reply src/gulcalc/gulcalc.cpp:890]: That's outside of the for-loop though. It ends at 824

Comment (ndaqjn) [reply src/gulcalc/gulcalc.cpp:890]: My bad, the diff view had hidden some code I missed

**Pull request merged**